controllers/pkg/reconcilers/bootstrap/reconciler.go

```diff
@@ -0,0 +1,277 @@
+						}
+
+						if err := clusterClient.Apply(ctx, cr); err != nil {
+							msg := fmt.Sprintf("cannot apply secret to cluster %s", clusterName)
+							r.l.Error(err, msg)
+							return ctrl.Result{RequeueAfter: 10 * time.Second}, errors.Wrap(err, msg)
+						}
+					}
+				}
+			}
+			if !found {
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+			}
+		}
+	} else {
+		// this branch handles manifest installation
+		cl, ok := cluster.Cluster{Client: r.Client}.GetClusterClient(cr)
+		if ok {
+			r.l.Info("reconcile")
+			clusterClient, ready, err := cl.GetClusterClient(ctx)
+			if err != nil {
+				msg := "cannot get clusterClient"
+				r.l.Error(err, msg)
+				return ctrl.Result{RequeueAfter: 30 * time.Second}, errors.Wrap(err, msg)
+			}
+			if !ready {
+				r.l.Info("cluster not ready")
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+			}
+
+			// install resources
+			resources, err := r.getResources(ctx, cl.GetClusterName())
+			if err != nil {
+				msg := "cannot get resources"
+				r.l.Error(err, msg)
+				return ctrl.Result{RequeueAfter: 10 * time.Second}, errors.Wrap(err, msg)
+			}
+			for _, resource := range resources {
+				r.l.Info("install manifest", "resources",
+					fmt.Sprintf("%s.%s.%s", resource.GetAPIVersion(), resource.GetKind(), resource.GetName()))
+				if err := clusterClient.Apply(ctx, &resource); err != nil {
+					r.l.Error(err, "cannot apply resource to cluster", "name", resource.GetName())
+				}
+			}
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) getResources(ctx context.Context, clusterName string) ([]unstructured.Unstructured, error) {
+	repos := &porchconfigv1alpha1.RepositoryList{}
+	if err := r.porchClient.List(ctx, repos); err != nil {
+		return nil, err
+	}
+
+	stagingRepoName := ""
+	for _, repo := range repos.Items {
+		if _, ok := repo.Annotations["nephio.org/staging"]; ok {
+			stagingRepoName = repo.GetName()
+		}
+	}
+
+	prList := &porchv1alpha1.PackageRevisionList{}
+	if err := r.porchClient.List(ctx, prList); err != nil {
+		return nil, err
+	}
+
+	prKeys := []types.NamespacedName{}
+	for _, pr := range prList.Items {
+		if pr.Spec.RepositoryName == stagingRepoName && pr.Annotations["nephio.org/site"] == clusterName {
+			prKeys = append(prKeys, types.NamespacedName{Name: pr.GetName(), Namespace: pr.GetNamespace()})
+		}
+	}
+	resources := []unstructured.Unstructured{}
+	for _, prKey := range prKeys {
+		prr := &porchv1alpha1.PackageRevisionResources{}
+		if err := r.porchClient.Get(ctx, prKey, prr); err != nil {
+			r.l.Error(err, "cannot get package resvision resourcelist", "key", prKey)
+			return nil, err
+		}
+
+		res, err := r.getResourcesPRR(prr.Spec.Resources)
+		if err != nil {
+			r.l.Error(err, "cannot get resources", "key", prKey)
+			return nil, err
+		}
+		resources = append(resources, res...)
+	}
+
+	return resources, nil
+}
+
+func includeFile(path string, match []string) bool {
```

Comment:
Soon we need to build some util libs for controllers that do this, I do the same in PV, and I know we did it in other places too. Not this PR, not yet...

Reply:
yes this is reused all over the place. Even in porch. So we should build some libs to interact with porch.

Reply:
exactly, I want to do that in porch/controllers/pkg I think.